rfd/0086/mariposa.md

```diff
@@ -0,0 +1,97 @@
+
+### Telemetry endpoint
+
+##### `GetStatus GET /status`
+
+This API will expose the state of all services associated with this ContainerPilot instance, including their dependencies.
+
+*Example Response*
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Content-Length: 328
+{
+  "services": [
+    {
+      "name": "nginx",
+      "address": "192.168.1.100",
+      "port": 80,
+      "status": "healthy"
+      "depends_on": ["backend1", "backend2"]
+    },
+    {
+      "name": "consul_agent",
+      "address": "localhost",
+      "port": 8500,
+      "status": "nonadvertised"
+    }
+  ]
+}
+```
+
+### Control plane endpoint
+
+##### `PutEnv POST /v3/env`
+
+This API allows a hook to update the environment variables that ContainerPilot provides to lifecycle hooks. The body of the POST must be in JSON format. The keys will be used as the environment variable to set, and the values will be the values to set for those environment variables. The environment variables take effect for all future processes spawned and override any existing environment variables. Unsetting a variable is supported by passing an empty string or `null` as the JSON value for that key. This API returns HTTP400 if the key is not a valid environment variable name, otherwise HTTP200 with no body.
+
+*Example Request*
+
+```
+curl -XPOST \
+    -d '{"ENV1": "value1", "ENV2": "value2", "ENV_TO_CLEAR": ""}' \
+    --unix-socket /var/containerpilot.sock \
+    http:/v3/env
+```
+
+##### `PutMetric POST /v3/metric`
+
+This API allows a sensor hook to update Prometheus metrics. (This allows sensor hooks to do so without having to suppress their own logging, which is required under 2.x.) The body of the POST must be in JSON format. The keys will be used as the metric names to update, and the values will be the values to set/add for those metrics. The API will return HTTP400 if the metric is not one that ContainerPilot is configuring, otherwise HTTP200 with no body.
+
+*Example Request*
+
+```
+curl -XPOST \
+    -d '{"my_counter_metric": 2, "my_gauge_metric": 42.42}' \
+    --unix-socket /var/containerpilot.sock \
+    http:/v3/metric
+```
+
+##### `Reload POST /v3/reload`
+
+This API allows a hook to force ContainerPilot to reload its configuration from file. This replaces the SIGHUP handler from 2.x and behaves identically: all pollables are stopped, the configuration file is reloaded, and the pollables are restarted without interfering with the services. This endpoint returns a HTTP200 with no body.
+
+*Example Request*
+
+```
+curl -XPOST \
+    --unix-socket /var/containerpilot.sock \
+    http:/v3/reload
+```
+
+##### `MaintenanceMode POST /v3/maintenance`
```

Comment:
Perhaps instead of a blind toggle we shold require the intent to be passed? If a blind toggle then you really need to query the current state before POSTing, and then that introduces timing issues. So maybe we pass `enable` or `disable` in the request body... or just make a `POST` to `/v3/maintenance/{enable,disable}` and just support `GET` on `/v3/maintenance` in case one needs to see the current state.